File: OrchidCore/src/main/java/com/eden/orchid/impl/compilers/parsers/CSVParser.java
package com.eden.orchid.impl.compilers.parsers;

import com.eden.orchid.api.compilers.OrchidParser;
import com.univocity.parsers.csv.CsvParser;
import com.univocity.parsers.csv.CsvParserSettings;
import com.univocity.parsers.tsv.TsvParser;
import com.univocity.parsers.tsv.TsvParserSettings;
import org.apache.commons.io.IOUtils;

import javax.inject.Inject;
import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public final class CSVParser extends OrchidParser {

    @Inject
    public CSVParser() {
        super(100);
    }

    @Override
    public String[] getSourceExtensions() {
        return new String[] {"csv", "tsv"};
    }

    @Override
    public Map<String, Object> parse(String extension, String input) {
        List<String[]> allRows;

        if(extension.equalsIgnoreCase("csv")) {
            CsvParserSettings settings = new CsvParserSettings();
            settings.getFormat().setLineSeparator("\n");
            CsvParser parser = new CsvParser(settings);

            allRows = parser.parseAll(IOUtils.toInputStream(input, Charset.forName("UTF-8")));
        }
        else {
            TsvParserSettings settings = new TsvParserSettings();
            settings.getFormat().setLineSeparator("\n");
            TsvParser parser = new TsvParser(settings);

            allRows = parser.parseAll(IOUtils.toInputStream(input, Charset.forName("UTF-8")));
        }

        List<Map<String, Object>> array = new ArrayList<>();

        String[] cols = allRows.get(0);

        for (int i = 1; i < allRows.size(); i++) {
            Map<String, Object> object = new HashMap<>();

            for (int j = 0; j < cols.length; j++) {
                object.put(cols[j], allRows.get(i)[j]);
            }

            array.add(object);
        }

        Map<String, Object> object = new HashMap<>();
        object.put(OrchidParser.arrayAsObjectKey, array);
        return object;
    }

    @Override
    public String serialize(String extension, Object input) {
        return "";
    }
}

File: OrchidCore/src/main/kotlin/com/eden/orchid/impl/compilers/parsers/CSVParser.kt
package com.eden.orchid.impl.compilers.parsers

import com.eden.orchid.api.compilers.OrchidParser
import com.eden.orchid.api.options.annotations.Archetype
import com.eden.orchid.api.options.archetypes.ConfigArchetype
import com.univocity.parsers.csv.CsvParser
import com.univocity.parsers.csv.CsvParserSettings
import com.univocity.parsers.tsv.TsvParser
import com.univocity.parsers.tsv.TsvParserSettings
import org.apache.commons.io.IOUtils
import java.nio.charset.StandardCharsets
import java.util.ArrayList
import java.util.HashMap
import javax.inject.Inject

@Archetype(value = ConfigArchetype::class, key = "services.parsers.csv")
class CSVParser @Inject
constructor() : OrchidParser() {

    override fun getSourceExtensions(): Array<String> {
        return arrayOf("csv", "tsv")
    }

    override fun parse(extension: String, input: String): Map<String, Any> {
        val allRows: List<Array<String>>

        if (extension.equals("csv", ignoreCase = true)) {
            val settings = CsvParserSettings()
            settings.format.setLineSeparator("\n")
            val parser = CsvParser(settings)

            allRows = parser.parseAll(IOUtils.toInputStream(input, StandardCharsets.UTF_8))
        } else {
            val settings = TsvParserSettings()
            settings.format.setLineSeparator("\n")
            val parser = TsvParser(settings)

            allRows = parser.parseAll(IOUtils.toInputStream(input, StandardCharsets.UTF_8))
        }

        val array = ArrayList<Map<String, Any>>()

        val cols = allRows[0]

        for (i in 1 until allRows.size) {
            val row = HashMap<String, Any>()

            for (j in cols.indices) {
                row[cols[j]] = allRows[i][j]
            }

            array.add(row)
        }

        val rows = HashMap<String, Any>()
        rows[OrchidParser.arrayAsObjectKey] = array
        return rows
    }

    override fun serialize(extension: String, input: Any): String {
        return ""
    }
}
